### ISSUE TITLE
Watch stream dies without recovering

### ISSUE THREAD

**Issue by psontag** [nolar/kopf#787]
## Long story short
The Watch stream for a custom resource fails and does not get recreated again. This prevents the operator from seeing any events for that resource.

## Description

We have noticed that sometimes our Operator does no longer react to any resource changes on the cluster. The reason for this seems to be that the Watcher stream of the operator for that resource died without being reestablished. We were able to track down the following log message:

```bash
Watcher for <our-resource>@none has failed: Cannot connect to host 172.20.8.1:443 ssl:default [Connection reset by peer]
```
Restarting the operator after this log messages appears fixes the problem. Unfortunately we were no able to reproduce this behaviour yet and it only happens rarely. We will try to gather more information on our end and update this issue.

We are already using the following server and client timeouts:
```python
    settings.watching.server_timeout = 120
    settings.watching.client_timeout = 150
    settings.watching.connect_timeout = 5
```
So I don't think this is related to #204.

We assumed that kopf automatically restarts any failed watches automatically. Is this only the cases for certain kind of errors?
Is there any way to verify that all watchers are running? We could then use this logic as a workaround in our Operator health probe.

## Environment

* Kopf version: 1.32.1
* Kubernetes version: v1.19.11
* Python version: 3.8
* OS/platform: Debian

**Comment by psontag** [comment]
I dug through the code base a bit. From the error message I assume the issue happens somewhere in the watching task that is created here:
https://github.com/nolar/kopf/blob/46186c48ac510f1c9533343c5b4d6267c9d1bd34/kopf/_core/reactor/orchestration.py#L237

Which means the guarded tasked failed with an uncaught exception https://github.com/nolar/kopf/blob/46186c48ac510f1c9533343c5b4d6267c9d1bd34/kopf/_cogs/aiokits/aiotasks.py#L76

So it seems like something in the watcher function does not get caught. The problem might come from this place where all objects are listed initially.
https://github.com/nolar/kopf/blob/46186c48ac510f1c9533343c5b4d6267c9d1bd34/kopf/_cogs/clients/watching.py#L159

Looking through the code I could not find a `try/except` block that would catch the connection error for the API call, so it would bubble up all the way to the `guarded_task` function. 
I am new to the code base so I might have missed something.

**Comment by psontag** [comment]
Here is the traceback for the exception:

```
Traceback (most recent call last):
  File "/opt/venv/lib/python3.8/site-packages/aiohttp/connector.py", line 969, in _wrap_create_connection
    return await self._loop.create_connection(*args, **kwargs)  # type: ignore  # noqa
  File "/opt/pyenv/versions/3.8.10/lib/python3.8/asyncio/base_events.py", line 1050, in create_connection
    transport, protocol = await self._create_connection_transport(
  File "/opt/pyenv/versions/3.8.10/lib/python3.8/asyncio/base_events.py", line 1080, in _create_connection_transport
    await waiter
  File "/opt/pyenv/versions/3.8.10/lib/python3.8/asyncio/selector_events.py", line 848, in _read_ready__data_received
    data = self._sock.recv(self.max_size)
ConnectionResetError: [Errno 104] Connection reset by peer

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/venv/lib/python3.8/site-packages/kopf/_cogs/aiokits/aiotasks.py", line 69, in guard
    await coro
  File "/opt/venv/lib/python3.8/site-packages/kopf/_core/reactor/queueing.py", line 184, in watcher
    async for raw_event in stream:
  File "/opt/venv/lib/python3.8/site-packages/kopf/_cogs/clients/watching.py", line 83, in infinite_watch
    async for raw_event in stream:
  File "/opt/venv/lib/python3.8/site-packages/kopf/_cogs/clients/watching.py", line 159, in continuous_watch
    items, resource_version = await fetching.list_objs_rv(resource=resource, namespace=namespace)
  File "/opt/venv/lib/python3.8/site-packages/kopf/_cogs/clients/auth.py", line 45, in wrapper
    return await fn(*args, **kwargs, context=context)
  File "/opt/venv/lib/python3.8/site-packages/kopf/_cogs/clients/fetching.py", line 30, in list_objs_rv
    rsp = await errors.parse_response(await context.session.get(url))
  File "/opt/venv/lib/python3.8/site-packages/aiohttp/client.py", line 520, in _request
    conn = await self._connector.connect(
  File "/opt/venv/lib/python3.8/site-packages/aiohttp/connector.py", line 535, in connect
    proto = await self._create_connection(req, traces, timeout)
  File "/opt/venv/lib/python3.8/site-packages/aiohttp/connector.py", line 892, in _create_connection
    _, proto = await self._create_direct_connection(req, traces, timeout)
  File "/opt/venv/lib/python3.8/site-packages/aiohttp/connector.py", line 1051, in _create_direct_connection
    raise last_exc
  File "/opt/venv/lib/python3.8/site-packages/aiohttp/connector.py", line 1020, in _create_direct_connection
    transp, proto = await self._wrap_create_connection(
  File "/opt/venv/lib/python3.8/site-packages/aiohttp/connector.py", line 975, in _wrap_create_connection
    raise client_error(req.connection_key, exc) from exc
aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host 172.20.8.1:443 ssl:default [Connection reset by peer]
```

Looks like the `fetching.list_objs_rv(resource=resource, namespace=namespace)` call is indeed the problem here.

**Comment by nolar** [comment]
@philipp-sontag-by Thanks for reporting and especially for the stack trace. That is helpful.

The watching itself is secured by these 2 places:

https://github.com/nolar/kopf/blob/1.32.1/kopf/_cogs/clients/watching.py#L263-L264

https://github.com/nolar/kopf/blob/1.32.1/kopf/_cogs/clients/watching.py#L70

Sadly, the initial listing part of the stream is not secured.

Right now, I am working on a feature (in a private branch) of retrying **ALL** API requests if they fail with server-side or networking errors: listing, patching, watching, etc. You can preview how it is going to work in #788 (just created).

That PR would probably fix this issue too.

**Comment by psontag** [comment]
Thanks for the answer.  #788 should definitely improve this situation.

But what would happen if all retries are exhausted for listing the objects? I assume the resulting exception would still cause the watch to be stopped without being restarted? Is this the intended behaviour? 

I would expect that when a watch dies it is either restarted (or at least it is tried to restart it) or the Operator fails completely and would then be restarted by Kubernetes.

**Comment by djsly** [comment]
Happy to see this thread ! we are getting affected the same way here, where watch operations just hangs until we restart the operator. 

this is seen in an a quite large AKS cluster.

**Comment by psontag** [comment]
@djsly Do you also see the mentioned log message? Otherwise your problem might also be related to #204, which could be fixed by configuring client and server timeouts.

**Comment by djsly** [comment]
I talked with @OmegaVVeapon and it seems that we encountered this #787 issue

**Comment by OmegaVVeapon** [comment]
Yeah... we're indeed experiencing the same issues.
Using Kopf version [1.29.2](https://github.com/OmegaVVeapon/kopf-k8s-sidecar/blob/main/Pipfile.lock#L129)

We usually experience problems during the **startup** of kopf, when the resource observer is being created.
See https://github.com/OmegaVVeapon/kopf-k8s-sidecar/issues/21.

I was under the impression that once we cleared that initialization, were we good forever (this had been the case so far).

However, yesterday we saw the watch get stuck **after** working for several days, here's that stacktrace:

```
06/08/2021 22:22:24.064 -0400,grafana-sc-dashboard,"[2021-06-09 02:22:24,064] kopf.reactor.queuein [WARNING ] Unprocessed streams left for [(configmaps.v1, 'd9155a11-73f1-11ea-82f7-4ae59ab90b24'), (configmaps.v1, 'd3fb7535-157b-4211-b64d-bf3a8a5606ba'), (configmaps.v1, '316faaea-35e1-4bf7-9322-cb0a9099e18c'), (configmaps.v1, 'bbbb6809-8bf9-45d3-b264-548c6397fbec'), (configmaps.v1, 'f47f556a-e046-4967-8a76-8133272f1388'), (configmaps.v1, '036c3119-bb5f-4b85-9ad0-ec5bada7c7a7'), (configmaps.v1, '50f59033-72a9-11ea-93a5-8ef6353d371e'), (configmaps.v1, '2ae112b0-7278-11ea-93a5-8ef6353d371e'), (configmaps.v1, '34d3e805-608c-4656-afce-56faa8c12819'), (configmaps.v1, '8dd3898f-f08e-4c98-9a88-8a3c5e8c74af'), (configmaps.v1, '6814bc90-909d-4698-b47a-85b054463d27')].",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"[2021-06-09 02:22:24,165] kopf.reactor.orchest [ERROR   ] Watcher for configmaps.v1@none has failed: Cannot connect to host 192.168.0.1:443 ssl:default [Connect call failed ('192.168.0.1', 443)]",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,Traceback (most recent call last):,monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/aiohttp/connector.py"", line 969, in _wrap_create_connection",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"    return await self._loop.create_connection(*args, **kwargs)  # type: ignore  # noqa",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/usr/local/lib/python3.8/asyncio/base_events.py"", line 1025, in create_connection",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,raise exceptions[0],monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/usr/local/lib/python3.8/asyncio/base_events.py"", line 1010, in create_connection",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,    sock = await self._connect_sock(,monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/usr/local/lib/python3.8/asyncio/base_events.py"", line 924, in _connect_sock",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"    await self.sock_connect(sock, address)",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/usr/local/lib/python3.8/asyncio/selector_events.py"", line 496, in sock_connect",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,    return await fut,monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/usr/local/lib/python3.8/asyncio/selector_events.py"", line 528, in _sock_connect_cb",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"    raise OSError(err, f'Connect call failed {address}')",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"ConnectionRefusedError: [Errno 111] Connect call failed ('192.168.0.1', 443)",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,The above exception was the direct cause of the following exception:,monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,Traceback (most recent call last):,monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/kopf/utilities/aiotasks.py"", line 69, in guard",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,    await coro,monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/kopf/reactor/queueing.py"", line 177, in watcher",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,    async for raw_event in stream:,monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/kopf/clients/watching.py"", line 77, in infinite_watch",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,    async for raw_event in stream:,monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/kopf/clients/watching.py"", line 153, in continuous_watch",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"    items, resource_version = await fetching.list_objs_rv(resource=resource, namespace=namespace)",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/kopf/clients/auth.py"", line 44, in wrapper",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"    return await fn(*args, **kwargs, context=context)",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/kopf/clients/fetching.py"", line 30, in list_objs_rv",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,    rsp = await errors.parse_response(await context.session.get(url)),monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/aiohttp/client.py"", line 520, in _request",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,    conn = await self._connector.connect(,monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/aiohttp/connector.py"", line 535, in connect",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"    proto = await self._create_connection(req, traces, timeout)",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/aiohttp/connector.py"", line 892, in _create_connection",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"    _, proto = await self._create_direct_connection(req, traces, timeout)",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/aiohttp/connector.py"", line 1051, in _create_direct_connection",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,    raise last_exc,monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/aiohttp/connector.py"", line 1020, in _create_direct_connection",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"    transp, proto = await self._wrap_create_connection(",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"File ""/.venv/lib/python3.8/site-packages/aiohttp/connector.py"", line 975, in _wrap_create_connection",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"    raise client_error(req.connection_key, exc) from exc",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
06/08/2021 22:22:24.464 -0400,grafana-sc-dashboard,"aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host 192.168.0.1:443 ssl:default [Connect call failed ('192.168.0.1', 443)]",monitoring.prometheus-operator-grafana-7bd7b5d4f6-j8p9z.grafana-sc-dashboard
```

which I believe is the same error as seen in this ticket...
These were the last logs we saw, after which point, no more updates were processed.

I have since upgraded to the latest Kopf `1.32.1` and have started doing tests with that in a branch.

**UDPATE:** Oh, and we do use [client and server timeouts](https://github.com/OmegaVVeapon/kopf-k8s-sidecar/blob/main/app/sidecar.py#L20). 
Not idea if the defaults are good but that's worked well for the most part for us...

**Comment by psontag** [comment]
Hey @nolar,
the retry and timeout changes in the current release candidate seemed to have helped with this issue. We haven't seen it now since we started using the release candidate ~2 weeks ago. So big thanks for that 👍 

But I still have one question:

Looking at the code that caused this issue
https://github.com/nolar/kopf/blob/308e7ff89b037013e960a674b94f356d3491ffe8/kopf/_cogs/clients/watching.py#L158
it still looks like it could in theory be possible that the stream dies after all timeouts are exhausted. Did I miss something here? Should we maybe add a guard to this?

**Comment by nolar** [comment]
This should be fixed in the freshly release `kopf>=1.33.0`. Documentation: https://kopf.readthedocs.io/en/stable/configuration/#retrying-of-api-errors. (Note: there is also one additional timeout setting added, which can be useful for you.)

The retries cover all internal API operations of the framework: watching, patching, peering touching, cluster scanning, etc — including the initial listing of the resources (before the watch-stream starts). The default config attempts the API request for approximately 1 minute — should be enough; can be configured.

I close this issue as presumably solved. If this is not the case, feel free to reopen it with additional details on what is left.

**Comment by psontag** [comment]
I think you misunderstood what I was asking. I agree that the timeouts/retries will help here. 

But my question was whether the underlying issue is also fixed. My understanding is that if an exception is raised in the `fetching.list_objs` call the watch stream dies and does not get restarted. Is this correct?

If this is the case the watch stream would still die when all retries are exhausted. I would then either expect the operator to crash hard (so that it gets restarted by kubernetes) or to keep trying to reestablish the watch. But I think neither is the case. Am I wrong here?

**Comment by nolar** [comment]
I see your point. You are right, I missed this aspect. I’ll double-check again.

**Comment by psontag** [comment]
We have seen this error pop up again with the following traceback:
```
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/kopf/_cogs/aiokits/aiotasks.py", line 69, in guard
    await coro
  File "/usr/local/lib/python3.8/site-packages/kopf/_core/reactor/queueing.py", line 184, in watcher
    async for raw_event in stream:
  File "/usr/local/lib/python3.8/site-packages/kopf/_cogs/clients/watching.py", line 82, in infinite_watch
    async for raw_event in stream:
  File "/usr/local/lib/python3.8/site-packages/kopf/_cogs/clients/watching.py", line 158, in continuous_watch
    objs, resource_version = await fetching.list_objs(
  File "/usr/local/lib/python3.8/site-packages/kopf/_cogs/clients/fetching.py", line 28, in list_objs
    rsp = await api.get(
  File "/usr/local/lib/python3.8/site-packages/kopf/_cogs/clients/api.py", line 111, in get
    response = await request(
  File "/usr/local/lib/python3.8/site-packages/kopf/_cogs/clients/auth.py", line 45, in wrapper
    return await fn(*args, **kwargs, context=context)
  File "/usr/local/lib/python3.8/site-packages/kopf/_cogs/clients/api.py", line 78, in request
    response = await context.session.request(
  File "/usr/local/lib/python3.8/site-packages/aiohttp/client.py", line 544, in _request
    await resp.start(conn)
  File "/usr/local/lib/python3.8/site-packages/aiohttp/client_reqrep.py", line 905, in start
    self._continue = None
  File "/usr/local/lib/python3.8/site-packages/aiohttp/helpers.py", line 656, in __exit__
    raise asyncio.TimeoutError from None
asyncio.exceptions.TimeoutError
```
The weird things is that we did not see any retry attempts of kopf in the logs at all. The only related message we have found is this retry error
```
Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'ProtocolError('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))': /apis/prefix/v1/resource
```

Do you have any idea how this could happen? 

By the way I would be happy to provide a PR that properly guards the `fetching.list_objs` call, but I would probably need some pointers from you.

**Comment by nolar** [comment]
_Several notes for my future self:_

There are 2 options when the API error retrying limit is exceeded:

* Escalate and fail the whole operator and let it be fixed/restarted by humans or by Kubernetes.
* Ignore the errors in listing and continue reconnecting again and again forever.

The 2nd option is implemented in #821 and is accepted as a better alternative to match the behaviour of other parts of the operator. 

Specifically, when an API error happens in patching a resource object, only that object is failed, others do continue normally. Hence, if an API error happens in watching a resource kind, only that kind should fail, not others and not the whole operator. Of all the environment errors, the operator generally fails only when the cluster scanning fails in the very beginning; or because of its internal errors — but never because of the resource errors.

Besides, restarting the whole operator is not useful to resolve the API error that caused this restart, but such a restart would generate an even higher load on the API.

So, the errors are better to be ignored — but not silently; they still log the error messages and stack traces, as the whole situation (the error) is abnormal and requires attention. But it does not exclude the failed resource kind from being operated and will recover as soon as the API is back to normal.
